File: hub/store/handshake.ts
/** Agent intro card — hub-readable profile, not the L1 capability connect handshake. */

export interface AgentHandshakeProfile {
  host?: string;
  address?: string;
  models?: string[];
  skills?: string[];
  ask_me_about?: string[];
  preferred_file_format?: string;
  other_tools?: string[];
  published_at?: string;
}

/** Client body for PUT /v1/agents/:id/handshake. */
export type HandshakeInput = {
  host?: string;
  address?: string;
  models?: string[];
  skills?: string[];
  ask_me_about?: string[];
  preferred_file_format?: string;
  other_tools?: string[];
};

const MAX_ITEMS = 12;
const MAX_LEN = 80;
const MAX_FORMAT = 40;

/** Tokens, keys, paths — never stored on a handshake card. */
export function looksLikeSecret(value: string): boolean {
  const t = value.trim();
  if (!t) return false;
  if (/^(sk-|sk_live_|sk_test_|ghp_|github_pat_|xox[baprs]-|Bearer\s)/i.test(t)) {
    return true;
  }
  if (/(api[_-]?key|secret|token|password|credential)\s*[:=]/i.test(t)) {
    return true;
  }
  if (/^(\/Users\/|\/home\/|~\/|file:\/\/|[A-Za-z]:\\)/.test(t)) return true;
  return false;
}

/** @deprecated use looksLikeSecret */
export const looksSecretOrPath = looksLikeSecret;

function cleanOne(value: unknown, maxLen: number): string | undefined {
  if (typeof value !== "string") return undefined;
  const t = value.trim().slice(0, maxLen);
  if (!t || looksLikeSecret(t)) return undefined;
  return t;
}

function cleanList(value: unknown): string[] | undefined {
  if (!Array.isArray(value)) return undefined;
  const out: string[] = [];
  const seen = new Set<string>();
  for (const item of value) {
    const t = cleanOne(item, MAX_LEN);
    if (!t) continue;
    const key = t.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(t);
    if (out.length >= MAX_ITEMS) break;
  }
  return out.length > 0 ? out : undefined;
}

/** Strip secrets/paths and cap size. Empty fields are omitted. */
export function sanitizeHandshake(
  input: HandshakeInput | Record<string, unknown>,
  publishedAt: string,
): AgentHandshakeProfile {
  const host = cleanOne(input.host, MAX_LEN);
  const address = cleanOne(input.address, MAX_LEN);
  const models = cleanList(input.models);
  const skills = cleanList(input.skills);
  const askMeAbout = cleanList(input.ask_me_about);
  const preferred = cleanOne(input.preferred_file_format, MAX_FORMAT);
  const otherTools = cleanList(input.other_tools);
  const card: AgentHandshakeProfile = { published_at: publishedAt };
  if (host) card.host = host;
  if (address) card.address = address;
  if (models) card.models = models;
  if (skills) card.skills = skills;
  if (askMeAbout) card.ask_me_about = askMeAbout;
  if (preferred) card.preferred_file_format = preferred;
  if (otherTools) card.other_tools = otherTools;
  return card;
}

export function isHandshakeProfile(
  value: unknown,
): value is AgentHandshakeProfile {
  return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

File: core/src/daemon/handshake.rs
//! Intro-card sanitizer for `publish_handshake`. Names only — no secrets/paths.

use crate::hub_client::HandshakeCard;

const MAX_LIST: usize = 12;
const MAX_ITEM: usize = 80;
const MAX_HOST: usize = 40;
const MAX_ADDRESS: usize = 120;
const MAX_FORMAT: usize = 80;

pub fn looks_secret_or_path(raw: &str) -> bool {
    let t = raw.trim();
    if t.is_empty() {
        return false;
    }
    let lower = t.to_ascii_lowercase();
    if lower.starts_with("sk-")
        || lower.starts_with("sk_live")
        || lower.starts_with("sk_test")
        || lower.starts_with("ghp_")
        || lower.starts_with("github_pat_")
        || lower.starts_with("xox")
    {
        return true;
    }
    if lower.contains("api_key=")
        || lower.contains("api-key=")
        || lower.contains("secret=")
        || lower.contains("token=")
        || lower.contains("password=")
        || lower.contains("bearer ")
    {
        return true;
    }
    t.starts_with("/Users/")
        || t.starts_with("/home/")
        || t.starts_with("~/")
        || t.starts_with("file://")
        || (t.len() >= 3 && t.as_bytes()[1] == b':' && t.as_bytes()[2] == b'\\')
}

fn clip_item(raw: &str, max: usize) -> Option<String> {
    let t = raw.split_whitespace().collect::<Vec<_>>().join(" ");
    if t.is_empty() || looks_secret_or_path(&t) {
        return None;
    }
    if t.chars().count() > max {
        Some(t.chars().take(max).collect())
    } else {
        Some(t)
    }
}

fn clip_list(items: &[String], max_item: usize) -> Vec<String> {
    let mut out = Vec::new();
    let mut seen = std::collections::BTreeSet::new();
    for item in items {
        let Some(v) = clip_item(item, max_item) else {
            continue;
        };
        let key = v.to_ascii_lowercase();
        if !seen.insert(key) {
            continue;
        }
        out.push(v);
        if out.len() >= MAX_LIST {
            break;
        }
    }
    out
}

pub fn host_display_name(slug: &str) -> &'static str {
    match slug.trim().to_ascii_lowercase().as_str() {
        "cursor" => "Cursor",
        "claude" => "Claude",
        "chatgpt" | "chatgpt-web" => "ChatGPT",
        "claude-web" => "Claude",
        other if other.contains("chatgpt") => "ChatGPT",
        other if other.contains("claude") => "Claude",
        _ => "AI host",
    }
}

pub fn sanitize_handshake(card: HandshakeCard) -> HandshakeCard {
    HandshakeCard {
        host: card.host.as_deref().and_then(|s| clip_item(s, MAX_HOST)),
        address: card.address.as_deref().and_then(|s| clip_item(s, MAX_ADDRESS)),
        models: clip_list(&card.models, MAX_ITEM),
        skills: clip_list(&card.skills, MAX_ITEM),
        ask_me_about: clip_list(&card.ask_me_about, MAX_ITEM),
        preferred_file_format: card
            .preferred_file_format
            .as_deref()
            .and_then(|s| clip_item(s, MAX_FORMAT)),
        other_tools: clip_list(&card.other_tools, MAX_ITEM),
        published_at: card.published_at,
    }
}

pub fn handshake_notes(card: &HandshakeCard) -> String {
    let mut lines = vec!["Handshake".to_string()];
    if let Some(h) = &card.host {
        lines.push(format!("Host: {h}"));
    }
    if let Some(a) = &card.address {
        lines.push(format!("Address: {a}"));
    }
    if !card.models.is_empty() {
        lines.push(format!("Models: {}", card.models.join(", ")));
    }
    if !card.skills.is_empty() {
        lines.push(format!("Skills: {}", card.skills.join(", ")));
    }
    if !card.ask_me_about.is_empty() {
        lines.push(format!("Ask me about: {}", card.ask_me_about.join(", ")));
    }
    if let Some(f) = &card.preferred_file_format {
        lines.push(format!("Preferred files: {f}"));
    }
    if !card.other_tools.is_empty() {
        lines.push(format!("Other tools: {}", card.other_tools.join(", ")));
    }
    lines.join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn strips_secrets_and_paths() {
        let card = sanitize_handshake(HandshakeCard {
            host: Some("Cursor".into()),
            models: vec!["Composer".into(), "sk-live-secret".into(), "~/keys".into()],
            other_tools: vec!["github".into(), "Bearer abc.def".into()],
            ask_me_about: vec!["routing".into(), "api_key=foo".into()],
            preferred_file_format: Some("markdown".into()),
            ..Default::default()
        });
        assert_eq!(card.host.as_deref(), Some("Cursor"));
        assert_eq!(card.models, vec!["Composer"]);
        assert_eq!(card.other_tools, vec!["github"]);
        assert_eq!(card.ask_me_about, vec!["routing"]);
    }
}
